firebase-firestore/src/main/java/com/google/firebase/firestore/util/Executors.java

```diff
@@ -29,6 +40,39 @@
   /** An executor that executes the provided runnable immediately on the current thread. */
   public static final Executor DIRECT_EXECUTOR = Runnable::run;
 
+  /**
+   * An executor that runs tasks in parallel on Android's AsyncTask.THREAD_POOL_EXECUTOR.
+   *
+   * <p>Unlike the main THREAD_POOL_EXECUTOR, this executor manages its own queue of tasks and can
+   * handle an unbounded number of pending tasks.
+   */
+  public static final Executor BACKGROUND_EXECUTOR =
+      new Executor() {
+        AtomicInteger activeRunnerCount = new AtomicInteger(0);
+        Queue<Runnable> pendingTasks = new ConcurrentLinkedQueue<>();
+
+        @Override
+        public void execute(Runnable command) {
+          pendingTasks.add(command);
+
+          if (activeRunnerCount.get() < MAXIMUM_CONCURRENT_BACKGROUND_TASKS) {
+            // Note that the runner count could temporarily exceed
+            // MAXIMUM_CONCURRENT_BACKGROUND_TASKS if this is code path was executed in parallel.
+            // While undesired, this would merely queue another task on THREAD_POOL_EXECUTOR,
+            // and we are unlikely to hit the 120 pending task limit.
+            activeRunnerCount.incrementAndGet();
+            AsyncTask.THREAD_POOL_EXECUTOR.execute(
```

Comment:
We still need to handle the fact that `AsyncTask.THREAD_POOL_EXECUTOR.execute` can reject adding this task.

We should implement a caller runs policy, rather than trying to block on the executor becoming available. i.e.

```java
// acquire permit
Runnable wrappedCommand = // thing that runs and then releases semaphore
try {
  AsyncTask.THREAD_POOL_EXECUTOR.execute(wrappedCommand);
} catch (RejectedExecutionException ignored) {
  wrappedCommand.run();
} 
```